### ISSUE TITLE
Failure to resync after uplink restored.

### ISSUE THREAD

**Issue by longnamegood** [FingerlessGlov3s/OPNsensePIAWireguard#39]
My apologies, I do not yet have the skill to debug python and work out if the code is the problem.  I can follow instructions.   I have a number of devices in my home network that are set to reboot every couple of days; it is not synchronized. I  have noticed that sometimes when my upstream router reboots at 2am, when I  connect after 6am to Opensense because I have no functioning Wireguard VPN running, there is , in the dashboard, the gateway is down, the interface has hundreds of i/o errors, and the last Wireguard handshake was about 2am. To me, this implies that it never recovered from losing the uplink signal. I have recreated the situation by manually issuing a reboot command to the router and watching the dashboard as the link goes down, the handshakes stop and never recover when the router returns to service.  All I have to do is issue a reboot of the Opensense box, and in a few minutes everything is back to normal operations.  I do not understand why  Wireguard does not recover from losing the external link to the internet, note, the switches do not power cycle, so the physical link is always up. Where  do i go? should I be looking at setting up a cron job to check if the handshake is really stale and do something? Or, something else.  I am new, but trying to figure out  what to do.  I am looking at the code , but, I can not see where it checks for an obviously stale handshake.

**Comment by FingerlessGlov3s** [comment]
If installed correctly, the script will automatically recreate a new connection to a new PIA WireGuard server, long as the cron job is setup.

Have you got a cron job created like the below screenshot? As that monitors the handshake and makes it reconnect, when handshake is lost.
![image](https://github.com/FingerlessGlov3s/OPNsensePIAWireguard/assets/3495647/3ccf5c0b-0941-4989-921d-6c8aeace47fc)

**Comment by longnamegood** [comment]
Yes, I have those exact settings.
![image](https://github.com/FingerlessGlov3s/OPNsensePIAWireguard/assets/145184488/473169fb-c0fd-43c6-9139-544bf601e3b0)

**Comment by FingerlessGlov3s** [comment]
Can you please SSH the OPNsense Firewall, then run the below command, when the issue happens and report back 😊
```/conf/PIAWireguard.py debug```

**Comment by longnamegood** [comment]
I hope I got the information you need.  I removed the unique data about my install, except local IP addressing.
[teraterm-16sep23b-cleaned.log](https://github.com/FingerlessGlov3s/OPNsensePIAWireguard/files/12641378/teraterm-16sep23b-cleaned.log)

**Comment by FingerlessGlov3s** [comment]
Can you make sure your OPNsense is on 23.7.3 or later, then fetch the latest version of the script with the below. Then get another log of the debug output. I think it could be a DNS issue, but can't confirm until you've defo got the latest script.

The UUID bits are unique and are not sensitive information, 

```
fetch -o /conf https://raw.githubusercontent.com/FingerlessGlov3s/OPNsensePIAWireguard/main/PIAWireguard.py
```

**Comment by longnamegood** [comment]
I think you may be right.
[teraterm-16sep23c.log](https://github.com/FingerlessGlov3s/OPNsensePIAWireguard/files/12641436/teraterm-16sep23c.log)

**Comment by FingerlessGlov3s** [comment]
When it happens try the below commands, both commands should return "moved" texted. This will test if there's connectivity or just DNS playing up. If those commands don't return like my example, then the issue isn't with the script.

```
curl http://google.com
curl http://1.1.1.1
```

eg
```
root@firewall:~ # curl http://google.com
<HTML><HEAD><meta http-equiv="content-type" content="text/html;charset=utf-8">
<TITLE>301 Moved</TITLE></HEAD><BODY>
<H1>301 Moved</H1>
The document has moved
<A HREF="http://www.google.com/">here</A>.
</BODY></HTML>
root@firewall:~ # curl http://1.1.1.1
<html>
<head><title>301 Moved Permanently</title></head>
<body>
<center><h1>301 Moved Permanently</h1></center>
<hr><center>cloudflare</center>
</body>
</html>
```

**Comment by longnamegood** [comment]
It looks like you are right.
[teraterm-16sep23d.log](https://github.com/FingerlessGlov3s/OPNsensePIAWireguard/files/12641478/teraterm-16sep23d.log)

**Comment by FingerlessGlov3s** [comment]
You'll have to workout why your DNS or networking is breaking when your issues happens.

If you can put the upstream router in modem mode, if it exists, that may help.

Other than that, not much I can do here.

**Comment by longnamegood** [comment]
Thank you, you have helped very much.  This is an opportunity for me to go learn about firewalls, something I have put off for a long time.  Thank you for your time and patience, Respect!

**Comment by FingerlessGlov3s** [comment]
Happy I could help 😊

**Comment by iShark5060** [comment]
I had a similar issue with my upstream router (isp provided).
Solved it by setting "Maximum TTL for RRsets and messages" and "TTL for Host Cache" entries to 60 sec.
(Unbound DNS -> Advanced)

Doesn't SOLVE the issue, but mitigates it.

**Comment by longnamegood** [comment]
Thanks @iShark5060 , tried it, but, that did not help in my case.  I am really thankful that @FingerlessGlov3s took me through the steps of debugging, which I was able to follow.  I think I have located the problem, after hours of poking around.  I had both the WG interface and WAN interface both as a  potential  "gateway candidate"; once I removed the WG gateway entry from being a candidate everything started working. And now,  it restarts the Wireguard connection as soon as the uplink is restored. I also modified in the script all the   "requests.get(piaServerList)"  to "requests.get(piaServerList, timeout=8)" so that the script would throw an error and let me know that it did not get the file as expected, instead of  waiting for a response from the network which would never come.  I still have to learn about firewalls, but, this was a good learning experience. Cheers!
